Add SPI for setting default value of TrackingPreventionEnabled flag on _WKWebsiteDataStoreConfiguration

szewai opened the pull request:
#### 048ee145e7ccd7397833ab11ab76ee74256bbc92
<pre>
Add SPI for setting default value of TrackingPreventionEnabled flag on _WKWebsiteDataStoreConfiguration
<a href="https://bugs.webkit.org/show_bug.cgi?id=270505">https://bugs.webkit.org/show_bug.cgi?id=270505</a>
<a href="https://rdar.apple.com/124053569">rdar://124053569</a>

Reviewed by Chris Dumez.

In current implementation, when client does not explicitly set TrackingPreventionEnabled flag on WebsiteDataStore, we
will use the default value, which is the TCC permission value. For clients that do not want to use the TCC permission
value by chance, add a new SPI for setting the default value at data store creation time.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration defaultTrackingPreventionEnabledOverride]):
(-[_WKWebsiteDataStoreConfiguration setDefaultTrackingPreventionEnabledOverride:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::defaultTrackingPreventionEnabled const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::defaultTrackingPreventionEnabledOverride const):
(WebKit::WebsiteDataStoreConfiguration::setDefaultTrackingPreventionEnabledOverride):

Canonical link: <a href="https://commits.webkit.org/275753@main">https://commits.webkit.org/275753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fd5b1d95d8cf3686b53c73623954f3394b0c31d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42720 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21741 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45121 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45332 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38843 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19106 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35363 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18749 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36771 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16319 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/776 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38163 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46837 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17538 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14459 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42077 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40704 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9539 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19336 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/18802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
```diff
@@ -101,6 +101,7 @@ WK_CLASS_AVAILABLE(macos(10.13), ios(11.0))
 @property (nonatomic) _WKUnifiedOriginStorageLevel unifiedOriginStorageLevel WK_API_AVAILABLE(macos(13.3), ios(16.4));
 @property (nonatomic, nullable, copy) NSNumber *volumeCapacityOverride WK_API_AVAILABLE(macos(14.0), ios(17.0));
 @property (nonatomic) BOOL isDeclarativeWebPushEnabled WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+@property (nonatomic, nullable, copy) NSNumber *defaultTrackingPreventionEnabledOverride WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
Confused why this returns a NSNumber* and not a BOOL?

cdumez replied:
Oh I see, it is because it's an optional BOOL, so we want the `nil` state..

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
```diff
@@ -807,6 +807,24 @@ - (void)setResourceLoadStatisticsDebugModeEnabled:(BOOL)enabled
+
+- (void)setDefaultTrackingPreventionEnabledOverride:(NSNumber *)defaultTrackingPreventionEnabledOverride
+{
+    std::optional<bool> enabled = std::nullopt;
```
` = std::nullopt` is unnecessary.

cdumez reviewed (approved):
LGTM with nit.

cc @beidson in case he has a better proposal than `NSNumber *` in the SPI here.

szewai committed:
Add SPI for setting default value of TrackingPreventionEnabled flag on _WKWebsiteDataStoreConfiguration
https://bugs.webkit.org/show_bug.cgi?id=270505
rdar://124053569

Reviewed by Chris Dumez.

In current implementation, when client does not explicitly set TrackingPreventionEnabled flag on WebsiteDataStore, we
will use the default value, which is the TCC permission value. For clients that do not want to use the TCC permission
value by chance, add a new SPI for setting the default value at data store creation time.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration defaultTrackingPreventionEnabledOverride]):
(-[_WKWebsiteDataStoreConfiguration setDefaultTrackingPreventionEnabledOverride:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::defaultTrackingPreventionEnabled const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::defaultTrackingPreventionEnabledOverride const):
(WebKit::WebsiteDataStoreConfiguration::setDefaultTrackingPreventionEnabledOverride):

Canonical link: https://commits.webkit.org/275753@main
